### PR TITLE
Add documentation for firmware resiliency

### PR DESCRIPTION
**Pull request by sean-m-mcginn**
This change adds a section for firmware resiliency as well as updates existing sections that are related to firmware resiliency

Signed-off-by: Sean McGinn <sean.mcginn@intel.com>

### CODE REVIEW

File: source/security/firmware-resiliency-and-recovery.rst [hunk 1]
```diff
@@ -0,0 +1,96 @@
+.. _firmware-resiliency-and-recovery:
+
+Firmware Resiliency and Recovery
+--------------------------------
+
+|SPN| implements a resiliency and recovery mechanism for boot-critical firmware. This along with :ref:`firmware-update` make |SPN| compliant with NIST SP 800-193.
```
Comment (rprangar) [inline source/security/firmware-resiliency-and-recovery.rst:6]: Regarding NIST compliancy - please remove that.

File: source/security/firmware-resiliency-and-recovery.rst [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            |   EMTY   |  0x072000(0xFF272000)  |  0x21b000  |  Uncompressed,  NR    |
+            +------------------------------------------------------------------------+
+            |                              NON VOLATILE                              |
+            +------------------------------------------------------------------------+
+            |   RSVD   |  0x071000(0xFF271000)  |  0x001000  |  Uncompressed,  NV    |
+            |   EMTY   |  0x000000(0xFF200000)  |  0x071000  |  Uncompressed,  NV    |
+            +----------+------------------------+------------+-----------------------+
+
+Whenever a failure is detected on one partition, a boot is tried from the other partition. If successful, the failing partition is overwritten with the working partition. Then,
+a normal boot to OS in executed, cancelling out a firmware update if one in flight. Thus, this can occur during a normal boot, or a update-related boot. So, the primary partition
```
Comment (rprangar) [inline source/security/firmware-resiliency-and-recovery.rst:77]: a normal boot to OS *is* executed

Comment (rprangar) [reply source/security/firmware-resiliency-and-recovery.rst:77]: firmware update if one *is* in flight.

**Review by rprangar — APPROVED** [review]
LGTM

File: source/security/firmware-resiliency-and-recovery.rst [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            +------------------------------------------------------------------------+
+            |   RSVD   |  0x071000(0xFF271000)  |  0x001000  |  Uncompressed,  NV    |
+            |   EMTY   |  0x000000(0xFF200000)  |  0x071000  |  Uncompressed,  NV    |
+            +----------+------------------------+------------+-----------------------+
+
+Whenever a failure is detected on one partition, a boot is tried from the other partition. If successful, the failing partition is overwritten with the working partition. Then,
+a normal boot to OS is executed, cancelling out a firmware update if one is in flight. Thus, this can occur during a normal boot, or a update-related boot. So, the primary partition
+can be copied over to the backup partition and vice versa.
+
+The diagram below shows the boot flow when a boot failure occurs on the primary boot partition.
```
Comment (lsubash) [inline source/security/firmware-resiliency-and-recovery.rst:80]: We can add a block diagram for the resiliency feature and later go to the flow charts for detailed view.

File: source/security/firmware-update.rst [hunk 4]
```diff
@@ -3,55 +3,71 @@
 Firmware Update
 ------------------
 
-|SPN| implements a secure and power fail-safe firmware update mechanism.
+|SPN| implements a secure and fail-safe firmware update mechanism.
 
-  * |SPN| image is designed to contain redundant boot components and depends on hardware assisted boot partition switch to support fail-safe update.
```
Comment (lsubash) [inline source/security/firmware-update.rst:8]: The descriptions on left side are required.

File: source/security/firmware-update.rst [hunk 5]
```diff
@@ -60,17 +76,25 @@ Generating capsule
...
+  The capsule tool is capable of incorporating multiple firmware images into single capsule binary.
 
-  For components inside the container region <4 byte id for component inside container : 4 byte id container id> along with associated component binary should be provided as input to the tool. This format is required for components inside container region because there can be multiple container inside BIOS region
+  Each firmware is identified and included in the capsule image using a string. These strings are constructed as follows:
+
+  * For non-container components, simply use its 4 character string identifier <4 byte comp id>.
+
+  * For container components, use its 4 character container identifier along with its 4 character component identifier <4 byte id for component inside container : 4 byte id container id>.
+
+  These firmware identifier strings should be followed by the assoicated binary path.
```
Comment (lsubash) [inline source/security/firmware-update.rst:89]: Limitations of the tool using redundant and non-redundant blocks at time is not supported any more

File: source/security/firmware-update.rst [hunk 6]
```diff
@@ -329,7 +363,7 @@ Trigger Update From Shell
 
 During development, one can use shell command to manually test firmware update without relying on support in OS.
 
-1. Copy ``FwuImage.bin`` into root directory on FAT partition of a USB key
+1. Copy ``FwuImage.bin`` into device/directory identified by CAPSULE_INFO_CFG_DATA
```
Comment (lsubash) [inline source/security/firmware-update.rst:367]: Mention the default one used in CFG file

File: source/tools/index.rst [hunk 7]
```diff
@@ -10,4 +10,5 @@ Tools
    StitchTool
    GenContainer
    ConfigTools
-   GpioTool
\ No newline at end of file
+   GpioTool
+   UcodeUtility
```
Comment (lsubash) [inline source/tools/index.rst:14]: Add a tutorial section - Mentioned about the tools to corrupt image at static and at run time.

File: source/security/firmware-resiliency-and-recovery.rst [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+.. image:: /images/primary_boot_flow_for_resiliency.png
+
+The diagram below shows the boot flow when a boot failure occurs on the backup boot partition.
```
Comment (lsubash) [inline source/security/firmware-resiliency-and-recovery.rst:84]: These diagrams need to be modified. We should capture two cases.
- Failure occured in Fwupdate
- Failure in primary boot partition.

**Review by lsubash — CHANGES_REQUESTED** [review]

File: source/security/firmware-update.rst [hunk 9]
```diff
@@ -79,22 +103,24 @@ After gathering required firmware binaries, capsule image can be generated using
                             Output file for signed image
       -q, --quiet           without output messages or temp files
 
-  For example, the following command generates a capsule image (``FwuImage.bin``) containing an IFWI image (``sbl.bios.bin``), CSME image (``csme.bin``), CSME Firmware Update Driver (``csme_fw_update_driver.bin``) and container component TSN MAC address inside container IPFW (``tsnmacaddr.bin``) signed by key ``FirmwareUpdateTestKey_Priv_RSA2048.pem``::
```
Comment (lsubash) [inline source/security/firmware-update.rst:82]: add example command for container component update as it gets removed in combination with Bios.

File: source/tutorials/ex_corrupt_sbl_component.rst [hunk 10]
```diff
@@ -0,0 +1,52 @@
+.. _ExerciseCorruptSblComponent:
```
Comment (lsubash) [inline source/tutorials/ex_corrupt_sbl_component.rst:1]: Add a note for build PCD to enable this tool.

File: source/tutorials/ex_corrupt_sbl_component.rst [hunk 11]
```diff
@@ -0,0 +1,52 @@
+.. _ExerciseCorruptSblComponent:
+
+Exercise \\- \ Corrupt |SPN| Component
+--------------------------------------
+
+CorruptComponentUtility (static corruption tool)
+************************************************
+
+``CorruptComponentUtility.py`` corrupts an SBL component (e.g. an item from its flash map) in either
```
Comment (lsubash) [inline source/tutorials/ex_corrupt_sbl_component.rst:9]: do mention not to include this tool in production.

File: source/tutorials/ex_corrupt_sbl_component.rst (hunk 10 above)
Comment (lsubash) [reply source/tutorials/ex_corrupt_sbl_component.rst:1]: Thanks.

File: source/tutorials/ex_corrupt_sbl_component.rst (hunk 11 above)
Comment (lsubash) [reply source/tutorials/ex_corrupt_sbl_component.rst:9]: i am saying about shell tool.

**Review by lsubash — APPROVED** [review]
Looks good.

**Comment by lsubash** [comment]
@sean-m-mcginn 
Remove WIP tag.

**Review by kvparikh89 — CHANGES_REQUESTED** [review]
Requesting changes to prevent merge till ADL-N is upstreamed

**Review by kvparikh89 — APPROVED** [review]

**Pull request merged**